File: runtime/permissions.test.ts
import { describe, expect, test } from "bun:test";
import {
  isAllowedByManifest, supportStatus, PermissionDeniedError,
  type PermissionsManifest,
} from "./permissions";

const m = (active: boolean, allow: string[]): PermissionsManifest =>
  ({ platform: "macos", active, allow });

describe("isAllowedByManifest (verb semantics)", () => {
  test("inactive allows all", () => expect(isAllowedByManifest("tray", m(false, []))).toBe(true));
  test("bare grants verbs", () => expect(isAllowedByManifest("fs:read", m(true, ["fs"]))).toBe(true));
  test("verb is exact", () => expect(isAllowedByManifest("fs:write", m(true, ["fs:read"]))).toBe(false));
  test("missing manifest treated as inactive", () => expect(isAllowedByManifest("tray", undefined)).toBe(true));
});

describe("supportStatus (platform axis)", () => {
  test("tray unsupported on ios", () => expect(supportStatus("tray", "ios")).toBe("unsupported"));
  test("shortcuts unsupported on ios", () => expect(supportStatus("shortcuts", "ios")).toBe("unsupported"));
  test("clipboard supported on ios", () => expect(supportStatus("clipboard:read", "ios")).toBe("supported"));
  test("tray supported on macos", () => expect(supportStatus("tray", "macos")).toBe("supported"));
  test("verb resolves via its module", () => expect(supportStatus("menu", "ios")).toBe("unsupported"));
});

describe("PermissionDeniedError", () => {
  test("carries code + permission", () => {
    const e = new PermissionDeniedError("clipboard:write");
    expect(e.code).toBe("PERMISSION_DENIED");
    expect(e.permission).toBe("clipboard:write");
    expect(e.message).toContain("clipboard:write");
  });
});

File: native/nim/tests/permissions_test.nim
import ../permissions

# permissions.nim importc's zapp_build_permissions_json (the CLI-generated getter).
# Never CALLED here (every test drives permissionsResetAndLoad directly), but
# permissionsEnsureInit references it, so stub it to satisfy the link — same as
# callbacks_test.nim stubs zapp_dispatch_event_to_js.
proc zapp_build_permissions_json(): cstring {.exportc, cdecl.} =
  cstring("{\"platform\":\"macos\",\"active\":false,\"allow\":[]}")

proc test() =
  # inactive => allow all
  permissionsResetAndLoad("{\"platform\":\"macos\",\"active\":false,\"allow\":[]}")
  doAssert permissionsIsAllowed(cstring"clipboard:read")
  doAssert permissionsIsAllowed(cstring"tray")
  # active + empty => deny all
  permissionsResetAndLoad("{\"platform\":\"macos\",\"active\":true,\"allow\":[]}")
  doAssert not permissionsIsAllowed(cstring"clipboard:read")
  # bare module grants its verbs
  permissionsResetAndLoad("{\"platform\":\"macos\",\"active\":true,\"allow\":[\"clipboard\"]}")
  doAssert permissionsIsAllowed(cstring"clipboard:read")
  doAssert permissionsIsAllowed(cstring"clipboard:write")
  doAssert permissionsIsAllowed(cstring"clipboard")
  doAssert not permissionsIsAllowed(cstring"fs:read")
  # exact verb grant, no sibling/bare bleed
  permissionsResetAndLoad("{\"platform\":\"macos\",\"active\":true,\"allow\":[\"fs:read\",\"shell:open\"]}")
  doAssert permissionsIsAllowed(cstring"fs:read")
  doAssert not permissionsIsAllowed(cstring"fs:write")
  doAssert not permissionsIsAllowed(cstring"fs")
  doAssert permissionsIsAllowed(cstring"shell:open")
  doAssert not permissionsIsAllowed(cstring"shell:trash")
  # malformed => fail open (inactive)
  permissionsResetAndLoad("{not json")
  doAssert permissionsIsAllowed(cstring"clipboard")
  # permission_id_for_invoke mapping (router.zc:21-36 parity)
  doAssert $permission_id_for_invoke(cstring"__clipboard:readText") == "clipboard:read"
  doAssert $permission_id_for_invoke(cstring"__clipboard:has") == "clipboard:read"
  doAssert $permission_id_for_invoke(cstring"__clipboard:writeText") == "clipboard:write"
  doAssert $permission_id_for_invoke(cstring"__dialog:open") == "dialog"
  doAssert $permission_id_for_invoke(cstring"__notif:show") == "notifications"
  doAssert $permission_id_for_invoke(cstring"__shortcuts:register") == "shortcuts"
  doAssert $permission_id_for_invoke(cstring"__screen:list") == "screen"
  doAssert $permission_id_for_invoke(cstring"__window:create") == "window:create"
  doAssert $permission_id_for_invoke(cstring"greet") == ""
  # permissions_check delegates to isAllowed
  permissionsResetAndLoad("{\"platform\":\"macos\",\"active\":true,\"allow\":[\"clipboard\"]}")
  doAssert permissions_check(cstring"clipboard:read", cstring"Clipboard.read")
  doAssert not permissions_check(cstring"fs:read", cstring"fs")
  echo "permissions ok"
test()
